### PR TITLE
Availability: React Refactor Phase 1

### PR DESCRIPTION
**Pull request by nicktiberi**
Re: #760.

@banukutlu - I spent some time this week experimenting with refactoring the availability JS in React. This is not a complete refactor, but I took the main portion of the HTML generating jQuery code and converted them into React components. We could build on this if we decide to continue down this path.

The basic gist of what I did was to create a base `<Availability />` component, which handles rendering the availability/holdings table. Within the `<Availability />` component, there is a `<HoldingDetails />` component with 4 sub-components:
- CourseReserveDueDate
- LocationInfo
- MapScanLink
- PublicNote

The advantage of this is that it allows us to separate out the various bits of logic that control the availability (parsing SIRSI responses, rendering UI, etc.).

My main reason for choosing React over [ViewComponent](https://viewcomponent.org/) was because we are rendering up to 100 of these components on a page at once, and since it is done asynchronously, my understanding was that we would either have to make 100 requests to the Rails endpoint that generates the ViewComponent (sounds v bad), or make a request to some Rails endpoint that generates multiple ViewComponents, and then individually place them on the page (less bad). The React approach seems more straightforward, IMO, but then again this is just my opinion and my natural tendency. I tested the performance with 100 React components in the search results and it seemed totally fine, which was great news!

No rush to check this out, of course, since of course our main priority is call number browse, but I'm curious to see what you think! 🙂 

### Still To Do
- [x] incorporate JSX/React linting

### Maybe Later
- [ ] move even more of the jQuery UI code (showing/hiding buttons, etc) into React components
- [ ] consider adding [Jest](https://github.com/facebook/jest) and unit test this at the individual React component level
- [ ] incorporate TypeScript for better autocomplete/tooling and type safety/error checking

### CODE REVIEW

**Comment by nicktiberi** [comment]
@banukutlu - I added some additional React-specific linting and took care of a few issues that it complained about. We can add more if needed: https://github.com/yannickcr/eslint-plugin-react#list-of-supported-rules

Code Climate is complaining about a few things, but I'm not sure if we care about all of its complaints. I will look into the complaints in `location_info.jsx` - that can probably be simplified a bit more.

File: app/javascript/availability/components/course_reserve_due_date.jsx
Comment (awead) [inline app/javascript/availability/components/course_reserve_due_date.jsx:15]: Do we know if `holding.dueDate` is in GMT?

File: app/javascript/availability/components/location_info.jsx
Comment (awead) [inline app/javascript/availability/components/location_info.jsx:86]: I wonder if you can take this one step further and so something like:

``` js
if (availability.isIllLink(holding)) {
  return (
    <IllLink holding={holding} />
  )
}

if (availability.isArchivalThesis(holding)) {
  return (
    <IllLink holding={holding} />
    <AeonLink holding={holding} />
  )
  
if (availability.isArchivalMaterial(holding)) {
return (
    <AeonLink holding={holding} />
  )
}
```

File: app/javascript/availability/components/location_info.jsx
Comment (awead) [inline app/javascript/availability/components/location_info.jsx:88]: Your second argument is the same as "self", i.e. `availability`. Wouldn't the `mapLocation` function be able to call `allLocations` from within itself?

**Review by awead — CHANGES_REQUESTED** [review]
Overall this looks good. I like seeing the re-useablility of the different React components, and I think it's an improvement. I had a couple of suggestions, but keep in mind I don't React, so some of my comments may be completely off.

I do see a lot of coupling of availability with a holding. If it was looking at this from a Ruby perspective, I might think about an Availability class, or component, has having a holding with it, since it seems that most of the time we're having to pass a holding into one of the availability's functions. That's just an observation, and not a suggestion, so take that as you want.

File: app/javascript/availability/components/course_reserve_due_date.jsx
Comment (nicktiberi) [reply app/javascript/availability/components/course_reserve_due_date.jsx:15]: I believe it is, but I am not 100% certain. I had to add the `timeZone` param in a previous commit because the CI server was running in a different time zone than the catalog server, which was causing tests to fail on CI.

File: app/javascript/availability/components/location_info.jsx
Comment (nicktiberi) [reply app/javascript/availability/components/location_info.jsx:86]: @awead - good call. I was thinking of saving that for a second stage of refactor but maybe it makes sense to just do this now. There is some more ILL/AEON logic in `availability/index.js` that can then be moved into these components in a later stage of the refactor.

File: app/javascript/availability/components/location_info.jsx
Comment (nicktiberi) [reply app/javascript/availability/components/location_info.jsx:88]: Ha! Yeah, this was just a copy paste, and it seems the first argument to `mapLocation` is always `allLocations`. Will change that. Good catch!

**Comment by nicktiberi** [comment]
> Overall this looks good. I like seeing the re-useablility of the different React components, and I think it's an improvement. I had a couple of suggestions, but keep in mind I don't React, so some of my comments may be completely off.
> 
> I do see a lot of coupling of availability with a holding. If it was looking at this from a Ruby perspective, I might think about an Availability class, or component, has having a holding with it, since it seems that most of the time we're having to pass a holding into one of the availability's functions. That's just an observation, and not a suggestion, so take that as you want.

Yeah, that is a good point. We may be able to move some more functions out of the mega `availability/index.js` file into these components in a later stage of the refactor and eliminate some more of these `availability.doSomething(holding)` calls.

File: app/javascript/availability/components/location_info.jsx
Comment (awead) [reply app/javascript/availability/components/location_info.jsx:86]: Up to you. If you want to "tie it off" now and save it for later, that's ok by me.

File: app/javascript/availability/components/location_info.jsx
Comment (nicktiberi) [reply app/javascript/availability/components/location_info.jsx]: @awead - moved the `mapLocation` function into this component and also removed the unnecessary argument. Whittling away at the mega file little by little. 🙂 

If you and @banukutlu are ok with it, I'd like to handle the AEON/ILL stuff in a separate PR, since this one is getting kinda big already.

**Review by awead — APPROVED** [review]

**Comment by nicktiberi** [comment]
@banukutlu - this should be ready to go, but I'll wait for your stamp of approval. 🙂

**Comment by banukutlu** [comment]
> @banukutlu - this should be ready to go, but I'll wait for your stamp of approval. 🙂

cool 😎 , I will get to it after the LC review, hopefully sometime tomorrow :)

File: app/javascript/availability/components/location_info.jsx
Comment (banukutlu) [reply app/javascript/availability/components/location_info.jsx:90]: I was exactly thinking the same thing like it would be very nice to see `<IllLink holding={holding} />` and `<AeonLink holding={holding} />` components. I am totally fine for these to be the next phase of the refactor since as Nick mentioned this PR has already good amount of changes.

File: app/javascript/availability/components/holding_details.jsx
Comment (banukutlu) [inline app/javascript/availability/components/holding_details.jsx:28]: thanks @nicktiberi  for adding these :)

File: app/javascript/availability/components/availability.jsx
Comment (banukutlu) [inline app/javascript/availability/components/availability.jsx:42]: @nicktiberi would it worth it to make toggle more as a component?

Comment (nicktiberi) [reply app/javascript/availability/components/availability.jsx]: @banukutlu - added a `<ViewMoreButton />` component and did a few more small cleanups with passing props into components, so this should be ready to go! 😎

File: app/javascript/availability/components/view_more_button.jsx
Comment (banukutlu) [inline app/javascript/availability/components/view_more_button.jsx:1]: thanks @nicktiberi :)

File: app/javascript/availability/components/availability.jsx
Comment (banukutlu) [inline app/javascript/availability/components/availability.jsx:43]: 👌🏻

File: app/javascript/availability/components/availability.jsx
Comment (banukutlu) [reply app/javascript/availability/components/availability.jsx]: awesome, thanks 😎

**Pull request merged**
